### PR TITLE
feat(payment_link): add multiple custom css support in business level 

### PR DESCRIPTION
**Pull request by sahkal**
## Type of Change

- [x] Enhancement

## Description

Currently we support one theme config for payment link that can be set in business profile level, with this PR we will allow multiple payment link theme configuration in business profile.

### Additional Changes

- [x] This PR modifies the API contract

```shell
"payment_link_config": {
        "theme": "#8A0808",
        "logo": "https://i.pinimg.com/736x/4d/83/5c/4d835ca8aafbbb15f84d07d926fda473.jpg",
        "seller_name": "sahkal",
        "display_sdk_only": false,
        "business_specific_configs": {
            "key1": {
                "theme": "#3B845E",
                "logo": "https://i.pinimg.com/736x/4d/83/5c/4d835ca8aafbbb15f84d07d926fda473.jpg",
                "seller_name": "sahkal",
                "display_sdk_only": true
            },
            "key2": {
                "theme": "#B202FF",
                "logo": "https://i.pinimg.com/736x/4d/83/5c/4d835ca8aafbbb15f84d07d926fda473.jpg",
                "seller_name": "sahkal",
                "enabled_saved_payment_method":true
                // "display_sdk_only": true
            }
        }
    }
}
```

we are adding a new optional param that is `business_specific_configs` for multi theme configuration

## How did you test it?

Create Merchant account

```shell
curl --location 'http://localhost:8080/accounts' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data-raw '{
  "merchant_id": "merchant_1719479157",
  "locker_id": "m0010",
  "merchant_name": "NewAge Retailer",
  "merchant_details": {
    "primary_contact_person": "John Test",
    "primary_email": "JohnTest@test.com",
    "primary_phone": "sunt laborum",
    "secondary_contact_person": "John Test2",
    "secondary_email": "JohnTest2@test.com",
    "secondary_phone": "cillum do dolor id",
    "website": "www.example.com",
    "about_business": "Online Retail with a wide selection of organic products for North America",
    "address": {
      "line1": "1467",
      "line2": "Harrison Street",
      "line3": "Harrison Street",
      "city": "San Fransico",
      "state": "California",
      "zip": "94122",
      "country": "US"
    }
  },
  "webhook_details": {
    "webhook_version": "1.0.1",
    "webhook_username": "ekart_retail",
    "webhook_password": "password_ekart@123",
    "webhook_url":"https://eny91mlyw8bh.x.pipedream.net",
    "payment_created_enabled": true,
    "payment_succeeded_enabled": true,
    "payment_failed_enabled": true
  },
  "sub_merchants_enabled": false,
  "metadata": {
    "city": "NY",
    "unit": "245"
  },
  "primary_business_details": [
    {
      "country": "US",
      "business": "default"
    }
  ]
}'
```

Create Payment Connector
```shell
curl --location 'http://localhost:8080/account/merchant_1719470020/connectors' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '
{
    "connector_type": "fiz_operations",
    "connector_name": "stripe",
    "connector_account_details": {
        "auth_type": "HeaderKey",
        "api_key":  {stripe-api-key}
    },
    "test_mode": true,
    "disabled": false,
    "connector_webhook_details": {
        "merchant_secret": {stripe-merchant-secret}
    },
    
    "payment_methods_enabled": [
        {
            "payment_method": "card",
            "payment_method_types": [
                {
                    "payment_method_type": "credit",
                    "card_networks": [
                        "AmericanExpress",
                        "DinersClub",
                        "Discover",
                        "JCB",
                        "Mastercard",
                        "Visa"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "accepted_currencies": {
                        "type": "enable_only",
                        "list": [
                            "CZK",
                            "RON",
                            "TRY",
                            "USD",
                            "GBP",
                            "CAD",
                            "AUD",
                            "ARS",
                            "CHF",
                            "CNY",
                            "DKK",
                            "GHS",
                            "JPY",
                            "PLN",
                            "NGN",
                            "NOK",
                            "KES",
                            "SEK",
                            "NZD",
                            "ZAR",
                            "SGD",
                            "INR"
                        ]
                    }
                },
                {
                    "payment_method_type": "debit",
                    "card_networks": [
                        "AmericanExpress",
                        "DinersClub",
                        "Discover",
                        "JCB",
                        "Mastercard",
                        "Visa"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "accepted_currencies": {
                        "type": "enable_only",
                        "list": [
                            "CZK",
                            "RON",
                            "TRY",
                            "USD",
                            "GBP",
                            "CAD",
                            "AUD",
                            "ARS",
                            "CHF",
                            "CNY",
                            "DKK",
                            "GHS",
                            "JPY",
                            "PLN",
                            "NGN",
                            "NOK",
                            "KES",
                            "SEK",
                            "NZD",
                            "ZAR",
                            "SGD",
                            "INR"
                        ]
                    }
                },
                {
                    "payment_method_type": "credit",
                    "card_networks": [
                        "Mastercard",
                        "Visa"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "accepted_currencies": {
                        "type": "enable_only",
                        "list": [
                            "EUR"
                        ]
                    }
                },
                {
                    "payment_method_type": "debit",
                    "card_networks": [
                        "Mastercard",
                        "Visa"
                    ],
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "accepted_currencies": {
                        "type": "enable_only",
                        "list": [
                            "EUR"
                        ]
                    }
                }
            ]
        },
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "redirect_to_url",
                    "payment_method_type": "affirm"
                }
            ]
        },
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "redirect_to_url",
                    "payment_method_type": "afterpay_clearpay"
                }
            ]
        },
        {
            "payment_method": "bank_redirect",
            "payment_method_types": [
                {
                    "payment_method_type": "sofort",
                    "payment_experience": "redirect_to_url",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": false
                },
                {
                    "payment_method_type": "bancontact_card",
                    "payment_experience": "redirect_to_url",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": false
                },
                {
                    "payment_method_type": "ideal",
                    "payment_experience": "redirect_to_url",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": false
                },
                {
                    "payment_method_type": "blik",
                    "payment_experience": "redirect_to_url",
                    "card_networks": null,
                    "accepted_currencies": null,
                    "accepted_countries": null,
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        },
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "redirect_to_url",
                    "payment_method_type": "klarna"
                }
            ]
        },
        {
            "payment_method": "pay_later",
            "payment_method_types": [
                {
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true,
                    "payment_experience": "invoke_sdk_client",
                    "payment_method_type": "klarna"
                }
            ]
        },
        {
            "payment_method": "card",
            "payment_method_types": [
                {
                    "payment_method_type": "credit",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        },
        {
            "payment_method": "card",
            "payment_method_types": [
                {
                    "payment_method_type": "debit",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        },
        {
            "payment_method": "wallet",
            "payment_method_types": [
                {
                    "payment_method_type": "apple_pay",
                    "payment_experience": "invoke_sdk_client",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        },
        {
            "payment_method": "wallet",
            "payment_method_types": [
                {
                    "payment_method_type": "google_pay",
                    "payment_experience": "invoke_sdk_client",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                },
                {
                    "payment_method_type": "paypal",
                    "payment_experience": "redirect_to_url",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "installment_payment_enabled": true,
                    "recurring_enabled": true
                }
            ]
        }
    ],
    "metadata": {
        "google_pay": {
            "allowed_payment_methods": [
                {
                    "type": "CARD",
                    "parameters": {
                        "allowed_auth_methods": [
                            "PAN_ONLY",
                            "CRYPTOGRAM_3DS"
                        ],
                        "allowed_card_networks": [
                            "AMEX",
                            "DISCOVER",
                            "INTERAC",
                            "JCB",
                            "MASTERCARD",
                            "VISA"
                        ]
                    },
                    "tokenization_specification": {
                        "type": "PAYMENT_GATEWAY",
                        "parameters": {
                            "gateway": "example",
                            "gateway_merchant_id": {publishable_key}
                        }
                    }
                }
            ],
            "merchant_info": {
                "merchant_name": "Narayan Bhat"
            }
        },
        "apple_pay": {
            "session_token_data": {
                "initiative": "web",
                "certificate": {certificate},
                "display_name": "applepay",
                "certificate_keys": {certificate_keys},
                "initiative_context": "hyperswitch-sdk-test.netlify.app",
                "merchant_identifier": "merchant.com.stripe.sang"
            },
            "payment_request_data": {
                "label": "applepay pvt.ltd",
                "supported_networks": [
                    "visa",
                    "masterCard",
                    "amex",
                    "discover"
                ],
                "merchant_capabilities": [
                    "supports3DS"
                ]
            }
        }
    }
}
'
```

Create Api-key 

```shell
curl --location 'http://localhost:8080/api_keys/merchant_1719479670' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: test_admin' \
--data '{
  "name": "API Key 1",
  "description": null,
  "expiration": "2025-09-23T01:02:03.000Z"
}'
```

Now, 

Updating the default business profile with payment link config with `business_specific_configs`

```shell
curl --location 'http://localhost:8080/account/merchant_1719479670/business_profile/pro_yl0tPMSCxxJlndl0aMHz' \
--header 'Content-Type: application/json' \
--header 'api-key: test_admin' \
--data '{
"payment_link_config": {
        "theme": "#8A0808",
        "logo": "https://i.pinimg.com/736x/4d/83/5c/4d835ca8aafbbb15f84d07d926fda473.jpg",
        "seller_name": "sahkal",
        "display_sdk_only": false,
        "business_specific_configs": {
            "key1": {
                "theme": "#3B845E",
                "logo": "https://i.pinimg.com/736x/4d/83/5c/4d835ca8aafbbb15f84d07d926fda473.jpg",
                "seller_name": "sahkal",
                "display_sdk_only": true
            },
            "key2": {
                "theme": "#B202FF",
                "logo": "https://i.pinimg.com/736x/4d/83/5c/4d835ca8aafbbb15f84d07d926fda473.jpg",
                "seller_name": "sahkal",
                "enabled_saved_payment_method":true
                
            }
        }
    }
}'
```

Do Payments Create

Case 1: incase, `business_specific_configs` is present and key is not send in payment create request, config will be picked up from the default config.

```shell
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_DZb1asJFT5ywRneS0niZUSrkXAiwGopwoWDKETATbIJqvPm2hj58q1Yrpbhqdb6D' \
--data-raw '
{
    "amount": 6540,
    "currency": "USD",
     "payment_link": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 6540,
    "customer_id": "stripesavecard_1234",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "setup_future_usage": "on_session",
    "customer_acceptance": {
        "acceptance_type": "offline",
        "accepted_at": "1963-05-03T04:07:52.723Z",
        "online": {
            "ip_address": "127.0.0.1",
            "user_agent": "amet irure esse"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "routing": {
        "type": "single",
        "data": "stripe"
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "session_expiry": 2423545
}'
```
Screenshot 2024-06-27 at 2 53 29 PM

Case 2:incase, `business_specific_configs` is present and key is send in the request, then it should pick up config from the config mapped to the key in business profile.

```shell
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_DZb1asJFT5ywRneS0niZUSrkXAiwGopwoWDKETATbIJqvPm2hj58q1Yrpbhqdb6D' \
--data-raw '
{
    "amount": 6540,
    "currency": "USD",
     "payment_link": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 6540,
    "customer_id": "stripesavecard_1234",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "setup_future_usage": "on_session",
    "customer_acceptance": {
        "acceptance_type": "offline",
        "accepted_at": "1963-05-03T04:07:52.723Z",
        "online": {
            "ip_address": "127.0.0.1",
            "user_agent": "amet irure esse"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "routing": {
        "type": "single",
        "data": "stripe"
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "session_expiry": 2423545,
    "payment_link_config_id":"key1"
}
```
Screenshot 2024-06-27 at 2 56 23 PM

Case 3 :incase, `business_specific_configs` is present and invalid key is send in the request, then it should pick up default config from business profile.

```shell
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_DZb1asJFT5ywRneS0niZUSrkXAiwGopwoWDKETATbIJqvPm2hj58q1Yrpbhqdb6D' \
--data-raw '
{
    "amount": 6540,
    "currency": "USD",
     "payment_link": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 6540,
    "customer_id": "stripesavecard_1234",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "setup_future_usage": "on_session",
    "customer_acceptance": {
        "acceptance_type": "offline",
        "accepted_at": "1963-05-03T04:07:52.723Z",
        "online": {
            "ip_address": "127.0.0.1",
            "user_agent": "amet irure esse"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "routing": {
        "type": "single",
        "data": "stripe"
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "session_expiry": 2423545,
    "payment_link_config_id":"key234"
}
```
Screenshot 2024-06-27 at 3 00 36 PM

```
Note: Next test cases, can be tested based on creating new business profile, and doing feature and edge case testing.
```
## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/payment_link.rs [hunk 1]
```diff
@@ -121,9 +123,15 @@ pub async fn initiate_payment_link_flow(
         payment_intent.currency,
         payment_intent.client_secret.clone(),
     )?;
-    let amount = currency
-        .to_currency_base_unit(payment_intent.amount.get_amount_as_i64())
-        .change_context(errors::ApiErrorResponse::CurrencyConversionFailed)?;
+
+    let required_conversion_type = StringMajorUnitForConnector;
```
Comment (jarnura) [inline crates/router/src/core/payment_link.rs:127]: We should have a different type for this like `StringMajorUnit`

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payment_link.rs (hunk 1 above)
Comment (sahkal) [reply crates/router/src/core/payment_link.rs:127]: i cant use pub struct `StringMajorUnit`  since we are already using it for conversion type pub struct `StringMajorUnit(String)`

Comment (sahkal) [reply crates/router/src/core/payment_link.rs]: done, added `StringMajorUnitForCore` type to impl on!

**Review by jarnura — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]
